### PR TITLE
daemon: refactor job runner for test

### PR DESCRIPTION
**Pull request by FeelyChau**
Todo:

- [x]  Unit test for `JobRunner`

### CODE REVIEW

**Review by WenheLI — APPROVED** [review]
LGTM

File: packages/daemon/src/service/pipeline.ts [hunk 1]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  output: any;
+  dataset: any;
+  modelPath: string;
+  modelPlugin: PluginPackage;
+  dataProcess: PluginPackage;
+  datasetProcess: PluginPackage;
+}
+
+/**
+ * job runner
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:75]: We need more to describe what the JobRunner is.

File: packages/daemon/src/service/pipeline.ts [hunk 2]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+ */
+export class JobRunner {
+  opts: RunnerOptions;
+  constructor(
+    opts: RunnerOptions) {
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:80]: Why this new line?

File: packages/daemon/src/service/pipeline.ts [hunk 3]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+    this.opts = opts;
+  }
+  getParams(params: string | null, ...extra: object[]): object {
+    if (params == null) {
+      return Object.assign({}, ...extra);
+    } else {
+      return Object.assign(JSON.parse(params), ...extra);
+    }
+  }
+  verifyPlugin (name: string): void {
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:90]:
```suggestion
  assertPlugin (name: string): void {
```
"verify" does mean a bool would be returned.

File: packages/daemon/src/service/pipeline.ts [hunk 4]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+    if (!this.opts.plugins[name]) {
+      throw new TypeError(`"${name}" plugin is required`);
+    }
+  }
+  dispatchJobEvent(jobStatus: PipelineStatus, step?: PluginTypeI, stepAction?: 'start' | 'end') {
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:95]: No return type.

File: packages/daemon/src/service/pipeline.ts [hunk 5]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+    const jobEvent = new JobStatusChangeEvent(
+      jobStatus,
+      step,
+      stepAction
+    );
+    this.opts.tracer.dispatch(jobEvent);
+  }
+  async runPlugin(type: PluginTypeI, ...args: any[]) {
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:103]: Ditto.

File: packages/daemon/src/service/pipeline.ts [hunk 6]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+    const plugin = this.opts.plugins[type].plugin;
+    this.dispatchJobEvent(PipelineStatus.RUNNING, type, 'start');
+    const result = await this.opts.runnable.start(plugin, ...args);
+    this.dispatchJobEvent(PipelineStatus.RUNNING, type, 'end');
+    return result;
+  }
+
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:110]: Unnecessary empty line.

File: packages/daemon/src/service/pipeline.ts [hunk 7]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+      dataDir
+    }));
+  }
+  async runDataAccess(dataDir: string): Promise<any> {
+    this.verifyPlugin('dataAccess');
+    return this.runPlugin('dataAccess', this.getParams(this.opts.plugins.dataAccess.params, {
+      dataDir
+    }));
+  }
+
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:126]: Ditto.

File: packages/daemon/src/service/pipeline.ts [hunk 8]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  async runDatasetProcess(dataset: any): Promise<void> {
+    if (this.opts.plugins.datasetProcess) {
+      await this.runPlugin('datasetProcess', dataset, this.getParams(this.opts.plugins.datasetProcess.params));
+    }
+  }
+
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:132]: Ditto.

File: packages/daemon/src/service/pipeline.ts [hunk 9]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  async runDataProcess(dataset: any): Promise<void> {
+    if (this.opts.plugins.dataProcess) {
+      await this.runPlugin('dataProcess', dataset, this.getParams(this.opts.plugins.dataProcess.params));
+    }
+  }
+
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:138]: Ditto.

File: packages/daemon/src/service/pipeline.ts [hunk 10]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  async runModelDefine(dataset: any): Promise<ModelResult> {
+    return {
+      modelPlugin: this.opts.plugins.modelDefine.plugin,
+      model: await this.runPlugin('modelDefine', dataset, this.getParams(this.opts.plugins.modelDefine.params))
+    };
+  }
+
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:145]: Ditto.

File: packages/daemon/src/service/pipeline.ts [hunk 11]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  async runModelLoad(dataset: any, modelPath: string): Promise<ModelResult> {
+    return {
+      modelPlugin: this.opts.plugins.modelLoad.plugin,
+      model: await this.runPlugin('modelLoad', dataset, this.getParams(this.opts.plugins.modelLoad.params, {
+        // specify the recover path for model loader by default.
+        recoverPath: modelPath
+      }))
+    };
+  }
+
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:155]: Ditto.

File: packages/daemon/src/service/pipeline.ts [hunk 12]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  async runModelTrain(dataset: any, model: RunnableResponse, modelPath: string) {
+    return this.runPlugin('modelTrain', dataset, model, this.getParams(this.opts.plugins.modelTrain.params, {
+      modelPath
+    }));
+  }
+
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:161]: Ditto.

File: packages/daemon/src/service/pipeline.ts [hunk 13]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+    return this.runPlugin('modelEvaluate', dataset, model, this.getParams(this.opts.plugins.modelEvaluate.params, {
+      modelDir: modelPath
+    }));
+  }
+  /**
+   * Generate the output package for a given job.
+   * @param job the job model for output.
+   * @param opts the options to used for generating the output.
+   */
+  async generateOutput(job: JobEntity, opts: GenerateOptions) {
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:173]: The JobRunner should not handle generating output IMHO.

File: packages/daemon/src/service/pipeline.ts [hunk 14]
```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  pipeline: PipelineEntity;
+  plugins: Partial<Record<PluginTypeI, PluginInfo>>;
+  tracer: Tracer;
+  runnable: PluginRunnable;
+  datasetRoot: string;
+}
+
+interface ModelResult {
+  model: RunnableResponse;
+  modelPlugin: PluginPackage;
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:62]:
```suggestion
  plugin: PluginPackage;
```

**Review by yorkie — CHANGES_REQUESTED** [review]

File: packages/daemon/src/service/pipeline.ts (hunk 6 above)
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts:110]: What's the standard style we follow here?

Comment (yorkie) [reply packages/daemon/src/service/pipeline.ts:110]: You should keep a consistent style for your class.

File: packages/daemon/src/service/pipeline.ts (hunk 1 above)
Comment (yorkie) [reply packages/daemon/src/service/pipeline.ts:75]: This line could be removed.

File: packages/daemon/src/service/pipeline.ts [hunk 15]
```diff
@@ -48,6 +48,220 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+    } else {
+      return Object.assign(JSON.parse(params), ...extra);
+    }
+  }
+
+  /**
+   * check if the plugin is configured, throw error if not
+   * @param name plugin name
+   */
+  assertPlugin (name: string): void {
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:101]:
```suggestion
  assertPlugin(name: string): void {
```

File: packages/daemon/src/service/pipeline.ts [hunk 16]
```diff
@@ -48,6 +48,220 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+    return this.runPlugin('modelEvaluate', dataset, model, this.getParams(this.opts.plugins.modelEvaluate.params, {
+      modelDir: modelPath
+    }));
+  }
+
+  /**
+   * run the pipeline
+   */
+  async run(): Promise<RunnerResult> {
+    const dataDir = path.join(this.opts.datasetRoot, `${this.opts.plugins.dataCollect.plugin.name}@${this.opts.plugins.dataCollect.plugin.version}`);
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:236]: This line is too long.

File: packages/daemon/src/service/pipeline.ts [hunk 17]
```diff
@@ -48,6 +48,220 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  runnable: PluginRunnable;
+  datasetRoot: string;
+}
+
+interface ModelResult {
+  model: RunnableResponse;
+  plugin: PluginPackage;
+}
+
+interface RunnerResult {
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:65]:
```suggestion
interface JobResult {
```

File: packages/daemon/src/service/pipeline.ts [hunk 18]
```diff
@@ -48,6 +48,220 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  output: any;
+  dataset: any;
+  modelPath: string;
+  modelPlugin: PluginPackage;
+  dataProcess: PluginPackage;
+  datasetProcess: PluginPackage;
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:71]: Wrap the above 3 fields in `plugins`?

File: packages/daemon/src/service/pipeline.ts [hunk 19]
```diff
@@ -48,6 +48,220 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  datasetRoot: string;
+}
+
+interface ModelResult {
+  model: RunnableResponse;
+  plugin: PluginPackage;
+}
+
+interface RunnerResult {
+  output: any;
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:66]: What does the output mean?

File: packages/daemon/src/service/pipeline.ts (hunk 17 above)
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:65]: Add some description.

File: packages/daemon/src/service/pipeline.ts [hunk 20]
```diff
@@ -160,100 +375,83 @@ export class PipelineService {
     return plugins;
   }
 
-  async startJob(job: JobEntity, pipeline: PipelineEntity, plugins: Partial<Record<PluginTypeI, PluginInfo>>, tracer: Tracer): Promise<void> {
-    const runnable = await this.pluginManager.createRunnable(job.id, tracer);
-    // save the runnable object
-    this.runnableMap[job.id] = runnable;
+  /**
+   * Generate the output package for a given job.
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:379]:
```suggestion
   * Generate the output dist for a given job.
```

File: packages/daemon/src/service/pipeline.ts [hunk 21]
```diff
@@ -48,6 +48,220 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
...
+  modelPlugin: PluginPackage;
+  dataProcess: PluginPackage;
+  datasetProcess: PluginPackage;
+}
+
+/**
+ * job runner
+ * handle the pipeline running: check pipeline config, load and run plugins, dispatch pipeline running event.
+ */
+export class JobRunner {
```
Comment (yorkie) [inline packages/daemon/src/service/pipeline.ts:78]: Shall we move this to https://github.com/alibaba/pipcook/blob/8c43574df00d2fa59f7d0b15c00ca417af3641a6/packages/daemon/src/runner/helper.ts?

**Review by yorkie — CHANGES_REQUESTED** [review]

File: packages/daemon/src/service/pipeline.ts (hunk 13 above)
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts]: Yeah, it's better. Updated.

File: packages/daemon/src/service/pipeline.ts (hunk 19 above)
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts:66]: The output from evaluate plugin.

Comment (yorkie) [reply packages/daemon/src/service/pipeline.ts]: Please write it in source code.

File: packages/daemon/src/service/pipeline.ts (hunk 21 above)
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts]: how about move runner into `utils/job-runner.ts`?

File: packages/daemon/src/service/pipeline.ts (hunk 19 above)
Comment (FeelyChau) [reply packages/daemon/src/service/pipeline.ts]: Renamed to `evaluateOutput`.

**Review by yorkie — APPROVED** [review]
Looks good to me.

**Pull request merged**